Sources/NIOSSL/SwiftCrypto/NIOSSLSecureBytes.swift

```diff
@@ -0,0 +1,429 @@
+//
+// Copyright (c) 2017-2022 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+public struct NIOSSLSecureBytes {
+    @usableFromInline
+    var backing: Backing
+
+    @inlinable
+    public init() {
+        self = .init(count: 0)
+    }
+
+    @usableFromInline
+    init(count: Int) {
+        self.backing = NIOSSLSecureBytes.Backing.create(randomBytes: count)
+    }
+
+    init (bytes: [UInt8]) {
+        self.backing = Backing.create(bytes: bytes)
+    }
+    /// Allows initializing a SecureBytes object with a closure that will initialize the memory.
+    @usableFromInline
+    init(unsafeUninitializedCapacity: Int, initializingWith callback: (inout UnsafeMutableRawBufferPointer, inout Int) throws -> Void) rethrows {
+        self.backing = Backing.create(capacity: unsafeUninitializedCapacity)
+        try self.backing._withVeryUnsafeMutableBytes { veryUnsafePointer in
+            // As Array does, we want to truncate the initializing pointer to only have the requested size.
+            var veryUnsafePointer = UnsafeMutableRawBufferPointer(rebasing: veryUnsafePointer.prefix(unsafeUninitializedCapacity))
+            var initializedCount = 0
+            try callback(&veryUnsafePointer, &initializedCount)
+
+            self.backing.count = initializedCount
+        }
+    }
+}
+
+extension NIOSSLSecureBytes {
+    @inlinable
+    mutating func append<C: Collection>(_ data: C) where C.Element == UInt8 {
+        let requiredCapacity = self.count + data.count
+        if !isKnownUniquelyReferenced(&self.backing) || requiredCapacity > self.backing.capacity {
+            let newBacking = Backing.create(capacity: requiredCapacity)
+            newBacking._appendBytes(self.backing, inRange: 0..<self.count)
+            self.backing = newBacking
+        }
+        self.backing._appendBytes(data)
+    }
+
+    
+    mutating public func reserveCapacity(_ n: Int) {
+        if self.backing.capacity >= n {
+            return
+        }
+
+        let newBacking = Backing.create(capacity: n)
+        newBacking._appendBytes(self.backing, inRange: 0..<self.count)
+        self.backing = newBacking
+    }
+}
+
+// MARK: - Equatable conformance, constant-time
+extension NIOSSLSecureBytes: Equatable {
+    public static func == (lhs: NIOSSLSecureBytes, rhs: NIOSSLSecureBytes) -> Bool {
+        //return safeCompare(lhs, rhs)
+        lhs.backing.withUnsafeBytes { lhsPtr in
+            rhs.backing.withUnsafeBytes { rhsPtr in
+                constantTimeCompare(lhsPtr, rhsPtr)
+            }
+        }
+    }
+}
+
+// MARK: - Collection conformance
+extension NIOSSLSecureBytes: Collection {
+    
+    public struct Index {
```

Comment:
I'm inclined to want us to diverge from Crypto here a bit: while the opaque index seemed like a good idea at the time I don't think it's serving us here.

We can delete `struct Index` and use an `Int` as the `Index` directly. Our `startIndex` and `endIndex` stay the same values, they just stop being wrapped in `struct Index`. This will also let us remove our `Strideable` and `Hashable` conformances for `struct Index` and also remove `func index(after:)` and `func index(before:)`, and lets us collapse our `Collection`/`BidirectionalCollection`/`RandomAccessCollection` conformances into a single conformance to `RandomAccessCollection`.